docs/architecture/decisions/0003-split-terraform-state-into-separate-aws-cluster-and-kubernetes-resource-phases.md

```diff
@@ -0,0 +1,46 @@
+# 3. Split terraform state into separate AWS cluster and Kubernetes resource phases
+
+Date: 2021-08-11
+
+## Status
+
+Accepted
+
+## Context
+
+Our Terraform will require the use of at least the `aws` and `kubernetes` providers, with EKS infrastructure resources defined via `aws` and in-cluster resources and configuration defined via `kubernetes`. As the `kubernetes` provider requires arguments that are determined from AWS resource outputs, a naïve implementation would look something like this:
+
+```
+provider "aws" {
+  region = var.region
+}
+
+resource "aws_eks_cluster" "cluster" {
+    ...
+}
+
+provider "kubernetes" {
+  host                   = aws_eks_cluster.cluster.endpoint
+  cluster_ca_certificate = base64decode(aws_eks_cluster.cluster.certificate_authority.0.data)
+  token                  = aws_eks_cluster.cluster.token
+}
+```
+
+As Terraform providers are not fully handled as nodes in the TF plan graph it cannot be guaranteed that AWS resources will be fully created at the point that Terraform configures the `kubernetes` provider, which will lead to non-deterministic behaviour, race conditions, etc. Because of this, [Terraform providers should only use config values that are known ahead of time](https://www.terraform.io/docs/language/providers/configuration.html#provider-configuration-1).
+
+## Decision
+
+Split Terraform resources into two separate states and phases (actual names TBD):
+
+1. cluster-infrastructure - EKS cluster and associated AWS resources. No `kubernetes` provider should be defined.
+2. cluster-services - Kubernetes resources (e.g. ingress controllers) and in-cluster configuration (e.g. `aws-auth` ConfigMap), with the `kubernetes` provider configured via TF remote data sources, after `cluster-infrastructure` has been applied
```

Comment:
remote data sources, or remote state?

Personally, I think remote state would be nicer. Maybe you could avoid having the aws provider in the cluster-services state at all (instead reading all the stuff it needs from the cluster-infrastructure's state).

Reply:
Sorry yes, that was the intent - I've clarified that, with a nod towards the potential need to access AWS info via remote data sources in some cases if necessary.